### ISSUE TITLE
update-psresource fails to successfully update module due to incorrect tmp path

### ISSUE THREAD

**Issue by demdante** [PowerShell/PSResourceGet#323]
So, I have a powershell module that I have been writing and published to the Powershell gallery.  I already have an earlier version of the module installed on my machine, so I tried to use "Update-PSResource" to pull the newer version of the module from the powershell gallery and install it over the earlier version.  However, when I ran the command, I got the below error:

![2020-12-10 17_45_15-Administrator_ Windows PowerShell ISE](https://user-images.githubusercontent.com/64720591/101845556-e32c1a80-3b46-11eb-97be-a401295b8a84.png)

So, I tried navigating to the path indicated in the error using windows explorer to see if I could figure out why the cmdlet failed, and I found that the temporary path that the cmdlet creates when extracting the powershell module is not valid, but only when the powershell module version number has 4 parts to it (eg 1.4.67.2).  For affected powershell modules, when update-psresource is extracting the module to a temporary path, the folder the cmdlet creates which represents the module's version number has the last part of cut off (using the above example, the folder would be named "1.4.67"), but the cmdlet is trying to access the module as if the last portion had not been cut off, resulting in the error.  

![2020-12-10 21_48_45-poshbox](https://user-images.githubusercontent.com/64720591/101855838-0c0ada80-3b5c-11eb-9f53-a08e9b6ee4e4.png)

**Comment by SydneyhSmith** [comment]
Thanks @demdante this is a helpful bug report/issue for us to look closer into and resolve. Looks like this might be a good testing scenario for us to look closer at...since you have published this package on the PowerShell Gallery would you mind sharing the name of the package so we can make sure this issue is properly addressed? Thanks! 
Right now the Gallery doesnt enforce any versioning requirements however we do highly encourage SemVer and may look towards greater SemVer integration

**Comment by demdante** [comment]
@SydneyhSmith  So, in attempting to document a series of steps one can take to replicate the issue, I have stumbled across several other issues that are similar in nature and is most likely directly related to the issue I describe above.  Unfortunately, the module I was using has been updated, so, as we will see below, the issue can't be replicated now, as it would require installing an older version of the module to allow you to run update-psresource.  Lets consider the powershell module PSWindowsupdate.  As you can see, most of its release versions contain 4 digits:

![site](https://user-images.githubusercontent.com/64720591/104816821-6eb75c00-5815-11eb-8795-c28bfe174c56.png)


Knowing this information, lets assume, for whatever reason, I want to install the most out of date version of the module, version 2.0.0.0, on a system that does not currently have it installed.  To do this and display the result, we would run this code:

```powershell
Install-PSResource pswindowsupdate -Version 2.0.0.0 -Verbose
get-module pswindowsupdate -ListAvailable
```

If the command successfully worked, then powershell should display the module information for pswindowsupdate.  But we don't see anything.  PSwindowsupdate has failed to install.  

![2021-01-16 11_24_51-13 68 184 198 - Remote Desktop Connection](https://user-images.githubusercontent.com/64720591/104817184-84c61c00-5817-11eb-978a-9da26450cbf5.png)


Or, so it would seem.  This is honestly the strangest part.  If we navigate to the directory defined in PSModulePath that the install-psresource cmdlet installs powershell modules to (for me, and for most I would imagine, this directory is C:\Program Files\WindowsPowerShell\Modules), we can see that a PSwindowsupdate has actually installed, but not correctly it would seem.  Can you identify the problem?:

![2021-01-16 12_38_05-13 68 184 198 - Remote Desktop Connection](https://user-images.githubusercontent.com/64720591/104818796-b6dc7b80-5821-11eb-8379-1fb048a7e6b3.png)

The directory containing the actual module whose name should be the version of the module has been shortened from 4 digits to 3.  Indeed, if you take a look in the PSGetModuleInfo.xml file in the old PSwindowsupdate module install directory, the full module version is correctly defined with 4 digits.  This seems to explain why Get-module is not listing the module, as, if you try to import the module by specifying the full path to the module manifest file, you see this:

![2021-01-16 12_46_53-13 68 184 198 - Remote Desktop Connection](https://user-images.githubusercontent.com/64720591/104819044-f2c41080-5822-11eb-881e-15327c724882.png)

So, a couple of things to note.  The command succeeds if you leave out the "-version" parameter and just try to install the latest version of the module, as powershellget does correctly name the folder when run the command without specifying version to install using the "-version" parameter:

![2021-01-16 12_50_41-13 68 184 198 - Remote Desktop Connection](https://user-images.githubusercontent.com/64720591/104819153-81d12880-5823-11eb-9d49-a886d350712c.png)

However, simply leaving off the "-version" parameter and installing the latest version is not always guaranteed to work,  To replicate the issue described in the initial post, I created and uploaded a test module to the powershell gallery with the same versioning schema as the module which generated the error in the initial post named module.test.  The module's release version's are 4 digits and is represented by the below expression were "x" is the digit which increments for every release:
1.0.x.0
Again, I want to make sure that we are aware that this version schema is the same as the version schema for the module which produced the error in the initial post.  So, lets try to install the above test module, which I have pushed to the powershell gallery with the release version of 1.0.1.0, with the below commands:

```powershell
Install-PSResource module.test -Verbose
Get-Module module.test -ListAvailable
```

If we run the above code, again, we see nothing.  If we go back to the folder in which the module was installed, we see this is the version which powershellget has determined and named the folder holding the module:

![2021-01-16 13_20_16-13 68 184 198 - Remote Desktop Connection](https://user-images.githubusercontent.com/64720591/104819754-8a2b6280-5827-11eb-8313-da1cd6fed542.png)

So, this would seem to indicate that release versions with a trailing zero will result in powershellget failing to install and update the module.  Once again, I want to just make sure that we understand that the issue demonstrated here is the same that is seen in the initial post.

Also, just a side note, it would really help with the troubleshooting if the info returned when specifying verbosity was a bit more....existing, as this does not help much when trying to troubleshoot:

![2021-01-16 11_18_33-13 68 184 198 - Remote Desktop Connection](https://user-images.githubusercontent.com/64720591/104817026-8fcc7c80-5816-11eb-8271-b093de7ed7ae.png)

Anyway, now, lets consider the SpeculationControl powershell module.  This module's re;lease version's consist of three digits rather than PSWindowsupdate's 4 digit release versions. 

![site2](https://user-images.githubusercontent.com/64720591/104817098-0b2e2e00-5817-11eb-804a-3fd189b96861.png)

Now, as we did with PSwindowsupdate, lets assume that I want to install an older version of the module.  To do so and display the results, we run the below commands:

```powershell
Install-PSResource SpeculationControl -Version 1.0.4 -Verbose
Get-Module SpeculationControl -ListAvailable
```

This is what we see when we run the above code:

![2021-01-16 11_28_24-13 68 184 198 - Remote Desktop Connection](https://user-images.githubusercontent.com/64720591/104817242-f605cf00-5817-11eb-9c1a-b9b82979796b.png)

So, as we can see, the older version of the SpeculationControl module has successfully installed, unlike the older version of the PSWindowsupdate module.  I am confident that this issue, where install-psresource fails to install the older version of PSWindowsupdate (whose release version, 2.0.0.0, contains 4 digits) correctly, but succeeds in installing the older version of SpeculationControl (whose release version, 1.0.4, contains 3 digits) is due to the same bug that I describe in my initial post due to the way in which the powershellget module is written and how both updatepsresource.cs and installpsresource.cs both initialize and manipulate an object defined in installhelper.cs.  Don't know if it was apparent, but I tried to identify the problematic code myself, I just haven't had the time to pinpoint the offending code.  I am not sure powershellget successfully determines the version number and correctly names the directory with 4 digits which will house the module when leaving out the "-version" parameter, but  cuts off part of the version number when specifying the version to install using the "-version" parameter in Install-psresource or when attempting to update a powershell module by running update-psresource.  There are so many different ways powershellget can fail that I don't think it can even be considered a finished product.  However, I think that this post now contains plenty of information to allow you to identify and resolve all the issues.  Feel free to let me know if you need more info.

**Comment by demdante** [comment]
@SydneyhSmith Does someone want to confirm that my post above provided the information that you needed?  Or a confirmation that it has even been read?  Anything?

**Comment by SydneyhSmith** [comment]
Thanks @demdante your response is great and super helpful...we do plan to fix this in the 3.0 release but just hadnt commented yet since we hadnt had a chance to address this issue yet, but will get to it as we do validation work the next couple of months...thanks again for your detailed information it is super helpful!

**Comment by alerickson** [comment]
This will be resolved in Preview 11.
